test/jdk/java/net/httpclient/ProxyServer.java

```diff
@@ -285,40 +309,110 @@ private boolean authorized(Credentials credentials,
+                            .findFirst()
+                            .map((s) -> s.substring("host: ".length()))
+                            .map(String::trim)
+                            .orElse(null);
                     // check authorization credentials, if required by the server
-                    if (credentials != null && !authorized(credentials, headers)) {
-                        String resp = "HTTP/1.1 407 Proxy Authentication Required\r\n" +
-                                      "Content-Length: 0\r\n" +
-                                      "Proxy-Authenticate: Basic realm=\"proxy realm\"\r\n\r\n";
-
-                        clientOut.write(resp.getBytes(UTF_8));
+                    if (credentials != null) {
+                        if (!authorized(credentials, headers)) {
+                            String resp = "HTTP/1.1 407 Proxy Authentication Required\r\n" +
+                                    "Content-Length: 0\r\n" +
+                                    "Proxy-Authenticate: Basic realm=\"proxy realm\"\r\n\r\n";
+                            clientSocket.setOption(StandardSocketOptions.TCP_NODELAY, true);
+                            clientSocket.setOption(StandardSocketOptions.SO_LINGER, 5);
+                            var buffer = ByteBuffer.wrap(resp.getBytes(ISO_8859_1));
+                            clientSocket.write(buffer);
+                            if (debug) {
+                                var linger = clientSocket.getOption(StandardSocketOptions.SO_LINGER);
+                                var nodelay = clientSocket.getOption(StandardSocketOptions.TCP_NODELAY);
+                                System.out.printf("Proxy: unauthorized, 407 sent back (%s/%s bytes); linger: %s, nodelay: %s%n",
+                                                  buffer.position(), buffer.limit(), linger, nodelay);
+                            }
+                            long drained = drain(clientSocket);
+                            if (debug) {
+                                System.out.printf("Proxy: drained: %s%n", drained);
+                            }
+                            clientSocket.shutdownOutput();
+                            try {
+                                // On windows, we additionally need to delay before
+                                // closing the socket. Otherwise we get a reset on the
+                                // client side (The connection was aborted by a software
+                                // on the host machine).
+                                // Delay 500ms before actually closing the socket
+                                if (isWindows()) Thread.sleep(500);
+                            } catch (InterruptedException x) {
+                                // OK
+                            }
+                            clientSocket.shutdownInput();
+                            close();
+                            return;
```

Comment:
I realise it isn't related to this change, but why is the test proxy closing the connection?

Reply:
There's no guarantee that the proxy will have read all the bytes sent by the client - even if it attempts to drain the connection. So the only sane reaction if you're not going to parse the request body is to close the connection.

Reply:
Right, I was mistaken. It actually is related to this change. You are testing what happens *if* the proxy closes the connection. Though that wouldn't be normal behavior for a proxy. If you are sending 407 to the client then you would want to keep the connection open so the client can retry the request. Maybe we need some comments in ProxyServer to indicate that the connection is being closed to test this specific scenario. Though if ProxyServer is used in other tests, I wonder if it might be better to use some flag or switch to enable this connection closing behavior?

Reply:
I have reworked the ProxyServer to keep reusing the connection if it can detect that there will be no request body. If there is a request body however, it will close the connection. That should leave things unchanged for tests that might have tried to use the ProxyServer for non tunneled GET/HEAD requests. I suspect that no test was using it for POST (in non tunnel mode) as that obviously could not work if authentication was required. That said, all tests are passing :-)

Reply:
Thanks. That looks fine now.